cmd/query/app/grpc_handler.go

```diff
@@ -35,21 +39,30 @@ const (
 	msgTraceNotFound = "trace not found"
 )
 
+var (
+	errGRPCMetricsQueryDisabled = status.Error(codes.Unimplemented, "metrics querying is currently disabled")
+	errMissingServiceNames      = status.Error(codes.InvalidArgument, "please provide at least one service name")
+	errMissingQuantile          = status.Error(codes.InvalidArgument, "please provide a quantile between (0, 1]")
+)
+
 // GRPCHandler implements the gRPC endpoint of the query service.
 type GRPCHandler struct {
-	queryService *querysvc.QueryService
-	logger       *zap.Logger
-	tracer       opentracing.Tracer
+	queryService        *querysvc.QueryService
+	metricsQueryService querysvc.MetricsQueryService
+	logger              *zap.Logger
+	tracer              opentracing.Tracer
+	clock               clock
 }
 
 // NewGRPCHandler returns a GRPCHandler
-func NewGRPCHandler(queryService *querysvc.QueryService, logger *zap.Logger, tracer opentracing.Tracer) *GRPCHandler {
+func NewGRPCHandler(queryService *querysvc.QueryService, metricsQueryService querysvc.MetricsQueryService, logger *zap.Logger, tracer opentracing.Tracer, clock clock) *GRPCHandler {
```

Comment:
is it worth adopting the functional options pattern here given the list of params is growing?

Reply:
Lately I've been trying to stay away from functional options and just use an options/params struct, it's less work to implement and has better discoverability.

Reply:
In this concrete example, we don't even need any additional struct, just remove this function and initialize GRPCHandler directly